moquette-0.16/broker/src/main/java/io/moquette/broker/security/ACLFileParser.java

```diff
@@ -0,0 +1,128 @@
+            return AuthorizationsCollector.emptyImmutableCollector();
+        }
+        if (!file.exists()) {
+            LOG.warn(
+                    String.format(
+                            "parsing not existing file %s, so fallback on default configuration!",
+                            file.getAbsolutePath()));
+            return AuthorizationsCollector.emptyImmutableCollector();
+        }
+        try {
+            Reader reader = Files.newBufferedReader(file.toPath(), UTF_8);
+            return parse(reader);
+        } catch (IOException fex) {
+            LOG.warn(
+                    String.format(
+                            "parsing not existing file %s, so fallback on default configuration!",
+                            file.getAbsolutePath()),
+                    fex);
+            return AuthorizationsCollector.emptyImmutableCollector();
+        }
+    }
+
+    /**
+     * Parse the ACL configuration file
+     *
+     * @param reader
+     *            to parse
+     * @return the collector of authorizations form reader passed into.
+     * @throws ParseException
+     *             if the format is not compliant.
+     */
+    public static AuthorizationsCollector parse(Reader reader) throws ParseException {
+        if (reader == null) {
+            // just log and return default properties
+            LOG.warn("parsing NULL reader, so fallback on default configuration!");
+            return AuthorizationsCollector.emptyImmutableCollector();
+        }
+
+        BufferedReader br = new BufferedReader(reader);
+        String line;
+        AuthorizationsCollector collector = new AuthorizationsCollector();
+
+        Pattern emptyLine = Pattern.compile("^\\s*$");
+        Pattern commentLine = Pattern.compile("^#.*"); // As spec, comment lines should start with '#'
+        Pattern invalidCommentLine = Pattern.compile("^\\s*#.*");
+        // This pattern has a dependency on filtering `commentLine`.
+        Pattern endLineComment = Pattern.compile("^([\\w\\s\\/\\+]+#?)(\\s*#.*)$");
+        Matcher endLineCommentMatcher;
+
+        try {
+            while ((line = br.readLine()) != null) {
+                if (line.isEmpty() || emptyLine.matcher(line).matches() || commentLine.matcher(line).matches()) {
+                    // skip it's a black line or comment
+                    continue;
+                } else if (invalidCommentLine.matcher(line).matches()) {
+                    // it's a malformed comment
+                    int commentMarker = line.indexOf('#');
+                    throw new ParseException(line, commentMarker);
+                }
+
+                endLineCommentMatcher = endLineComment.matcher(line);
+                if (endLineCommentMatcher.matches()) {
+                    line = endLineCommentMatcher.group(1);
+                }
+
+                collector.parse(line);
+            }
+        } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)